lms/services/lis_result_sourcedid.py

```diff
@@ -41,47 +42,50 @@ def fetch_students_by_assignment(
             .all()
         )
 
-    def upsert(self, lis_info, h_user, lti_user):
+    def upsert_from_request(self, request, h_user, lti_user):
```

Comment:
I'd like to have a chat about this, because I think there are some pros and cons. You definitely could hide all of that inside, and it's obviously tidier, but:

 * It's kind of hidden behaviour when reading the code, making it harder to follow the flow
 * It makes it less clear what data is involved
   * The function doesn't state it's intent quite as clearly
   * You could of course have a design where no functions accept or return any data, and it all just modifies global state
   * This is kind of horrid though
 * Means you can't pass a different things (e.g. in tests)
   * Although this also means you are now free to make a mistake you might not otherwise make

Reply:
This is true.

Part of what pyramid_services gives us is the ability for service factories to receive the request from pyramid_services, so I think we should give ourselves the permission to make use of that.

However, I do understand that it can also be a curse.

I think part of the solution to this is to be conservative about adding things to the request, given that the request is tantamount to a global object. We shouldn't add a new `request.foo` property or method lightly. But sometimes some things do qualify.

I think an argument can certainly be made for passing in `h_user` and `lti_user` for the sake of being explicit / not hiding things.

Reply:
> Means you can't pass a different things (e.g. in tests)

If we separated service factories from service `__init__()` methods then you could, so this particular one is easily fixable I think:

```python
class FooService:
    # Called by tests and anything else that wants to construct a FooService directly.
    def __init__(self, dep_1, dep_2, ...):
        ...

    # Registered with pyramid_services.
    @classmethod
    factory(cls, context, request):
        return cls(request.dep_1, request.dep_2, ...)
```